src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -922,3 +924,14 @@ fn use_extra_filename(bcx: &BuildContext<'_, '_>, unit: &Unit) -> bool {
     }
     true
 }
+
+/// Returns whether or not this unit should use a hash in the pkg_dir to make it unique.
+fn use_pkg_dir(unit: &Unit) -> bool {
+    if unit.mode.is_doc_test() || unit.mode.is_doc() {
+        // Doc tests do not have metadata.
+        false
+    } else {
+        // These always use metadata.
+        true
+    }
+}
```

Comment:
> I didn't touch docs because I've not fully investigated them yet.
On some playing around with this change outside of the new build dir
layout, I did see some breakages.

What kind of breakages did you see?

Reply:
The test failure had the appearance of the docs being cleared but not rebuilt.  It was the test from #16344.

Note that docs are having no change in behavior in this PR.  This PR is mostly unblocking resolving #8332.